File: src/mathproofmesh/inspiration/analogy_agent.py
from __future__ import annotations

from typing import Any, Iterable

from ..schemas import AnalogyMapping, NoveltySignature, ProblemContract, stable_hash
from .local_library import LocalAnalogyLibrary


class AnalogyAgent:
    """Map verified local structures; return empty rather than invent a precedent."""

    def __init__(self, library: LocalAnalogyLibrary, *, top_k: int = 6) -> None:
        self.library = library
        self.top_k = top_k

    def search(
        self,
        problem: ProblemContract,
        *,
        target_obligation_ids: Iterable[str],
        object_tags: Iterable[str] = (),
        operation_tags: Iterable[str] = (),
        mechanism_tags: Iterable[str] = (),
        graph_tags: Iterable[str] = (),
    ) -> list[AnalogyMapping]:
        records = self.library.search(
            query_text=problem.normalized_statement,
            object_tags=object_tags,
            operation_tags=operation_tags,
            mechanism_tags=mechanism_tags,
            graph_tags=graph_tags,
            top_k=self.top_k,
        )
        results: list[AnalogyMapping] = []
        for record in records:
            mapping = self._mapping_from_record(
                record,
                problem_hash=problem.integrity_hash,
                target_obligation_ids=list(target_obligation_ids),
            )
            if mapping is not None:
                results.append(mapping)
        return results

    @staticmethod
    def _mapping_from_record(
        record: dict[str, Any],
        *,
        problem_hash: str,
        target_obligation_ids: list[str],
    ) -> AnalogyMapping | None:
        objects = record.get("object_correspondence") or {}
        operations = record.get("operation_correspondence") or {}
        transferable = record.get("transferable_lemmas") or []
        non_transferable = record.get("non_transferable_conditions") or []
        risks = record.get("transfer_risks") or []
        if not all((objects, operations, transferable, non_transferable, risks)):
            return None
        signature = NoveltySignature(
            representation_tags=list(record.get("representation_tags", [])),
            mechanism_tags=[
                "structural_analogy",
                *list(record.get("mechanism_tags", [])),
            ],
            core_objects=list(objects.values()),
            key_transformations=list(operations.values()),
            proof_principles=list(record.get("proof_principles", [])),
            targeted_obligation_ids=target_obligation_ids,
        )
        digest = stable_hash(
            (record.get("record_id"), problem_hash, signature.normalized_hash)
        )
        return AnalogyMapping(
            analogy_id=f"analogy_{digest[:12]}",
            source_record_id=str(record["record_id"]),
            source_problem_summary=str(record.get("problem_summary", "")),
            target_problem_hash=problem_hash,
            object_correspondence={str(k): str(v) for k, v in objects.items()},
            operation_correspondence={str(k): str(v) for k, v in operations.items()},
            transferable_lemmas=[str(item) for item in transferable],
            non_transferable_conditions=[str(item) for item in non_transferable],
            transfer_risks=[str(item) for item in risks],
            required_bridge_lemmas=[
                str(item) for item in record.get("required_bridge_lemmas", [])
            ],
            novelty_signature=signature,
        )

File: mathproofmesh-core/src/main/java/io/github/aililuola/mathproofmesh/inspiration/AnalogyAgent.java
package io.github.aililuola.mathproofmesh.inspiration;

import io.github.aililuola.mathproofmesh.contract.AnalogyMapping;
import io.github.aililuola.mathproofmesh.contract.CanonicalJson;
import io.github.aililuola.mathproofmesh.contract.NoveltySignature;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/** Converts only verified local records into bounded analogy mappings. */
public final class AnalogyAgent {
  private final LocalAnalogyLibrary library;
  private final int topK;

  public AnalogyAgent(LocalAnalogyLibrary library, int topK) {
    this.library = java.util.Objects.requireNonNull(library, "library");
    if (topK <= 0) {
      throw new IllegalArgumentException("topK must be positive");
    }
    this.topK = topK;
  }

  public List<AnalogyMapping> search(
      String normalizedProblem,
      String problemHash,
      List<String> targetObligationIds,
      LocalAnalogyLibrary.Query query) {
    List<AnalogyMapping> result = new ArrayList<>();
    LocalAnalogyLibrary.Query effective =
        new LocalAnalogyLibrary.Query(
            normalizedProblem,
            query.objectTags(),
            query.operationTags(),
            query.mechanismTags(),
            query.graphTags(),
            query.obligationKinds(),
            query.mechanismChain(),
            query.graphMotifTags());
    for (LocalAnalogyLibrary.Record record : library.search(effective, problemHash, topK)) {
      mapping(record, problemHash, targetObligationIds).ifPresent(result::add);
    }
    return List.copyOf(result);
  }

  public java.util.Optional<AnalogyMapping> mapping(
      LocalAnalogyLibrary.Record record,
      String problemHash,
      List<String> targetObligationIds) {
    if (!record.verified()
        || record.objectCorrespondence().isEmpty()
        || record.operationCorrespondence().isEmpty()
        || record.transferableLemmas().isEmpty()
        || record.nonTransferableConditions().isEmpty()
        || record.transferRisks().isEmpty()) {
      return java.util.Optional.empty();
    }
    NoveltySignature signature =
        new NoveltySignature(
            new ArrayList<>(record.objectCorrespondence().values()),
            List.of(),
            new ArrayList<>(record.operationCorrespondence().values()),
            concat("structural_analogy", record.mechanismTags()),
            null,
            null,
            null,
            record.proofPrinciples(),
            Map.of(),
            record.representationTags(),
            targetObligationIds);
    String id =
        "analogy_"
            + CanonicalJson.stableHash(
                    List.of(record.recordId(), problemHash, signature.normalizedHash()))
                .substring(0, 12);
    return java.util.Optional.of(
        new AnalogyMapping(
            id,
            record.nonTransferableConditions(),
            signature,
            record.objectCorrespondence(),
            record.operationCorrespondence(),
            record.requiredBridgeLemmas(),
            record.problemSummary(),
            record.recordId(),
            problemHash,
            record.transferRisks(),
            record.transferableLemmas()));
  }

  private static List<String> concat(String first, List<String> rest) {
    List<String> result = new ArrayList<>();
    result.add(first);
    result.addAll(rest);
    return List.copyOf(result);
  }
}
